cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup_test.go

```diff
@@ -699,3 +699,150 @@ func TestNodeGroupMachineSetDeleteNodes(t *testing.T) {
 	}
 }
+
+func TestNodeGroupMachineSetDeleteNodesWithMismatchedNodeGroups(t *testing.T) {
+	nodes := make([]*apiv1.Node, 2)
+	machines := make([]*v1beta1.Machine, 2)
+	nodeObjects := make([]runtime.Object, 2)
+	machineObjects := make([]runtime.Object, 2)
+
+	machineSet1 := &v1beta1.MachineSet{
+		TypeMeta: v1.TypeMeta{
+			Kind: "MachineSet",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machineset1",
+			Namespace: "test-namespace",
+			UID:       "machineSet1",
+		},
+		Spec: v1beta1.MachineSetSpec{
+			Replicas: int32ptr(int32(len(machines) / 2)),
+		},
+	}
+
+	machineSet2 := &v1beta1.MachineSet{
+		TypeMeta: v1.TypeMeta{
+			Kind: "MachineSet",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machineset2",
+			Namespace: "test-namespace",
+			UID:       "machineSet2",
+		},
+		Spec: v1beta1.MachineSetSpec{
+			Replicas: int32ptr(int32(len(machines) / 2)),
+		},
+	}
+
+	for i := 0; i < len(nodes); i++ {
+		refMachineSet := machineSet1
+		if i > 0 {
+			refMachineSet = machineSet2
+		}
+
+		nodes[i] = &apiv1.Node{
+			TypeMeta: v1.TypeMeta{
+				Kind: "Node",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name: fmt.Sprintf("node-%d", i),
+				Annotations: map[string]string{
+					machineAnnotationKey: fmt.Sprintf("test-namespace/machine-%d", i),
+				},
+			},
+			Spec: apiv1.NodeSpec{
+				ProviderID: fmt.Sprintf("providerid-%d", i),
+			},
+		}
+
+		machines[i] = &v1beta1.Machine{
+			TypeMeta: v1.TypeMeta{
+				Kind: "Machine",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name:      fmt.Sprintf("machine-%d", i),
+				Namespace: "test-namespace",
+				OwnerReferences: []v1.OwnerReference{{
+					Name: refMachineSet.Name,
+					Kind: refMachineSet.Kind,
+					UID:  refMachineSet.UID,
+				}},
+			},
+			Status: v1beta1.MachineStatus{
+				NodeRef: &apiv1.ObjectReference{
+					Kind: nodes[i].Kind,
+					Name: nodes[i].Name,
+				},
+			},
+		}
+
+		nodeObjects[i] = nodes[i]
+		machineObjects[i] = machines[i]
+	}
+
+	controller, stop := mustCreateTestController(t, testControllerConfig{
+		nodeObjects:    nodeObjects,
+		machineObjects: append(machineObjects, machineSet1, machineSet2),
+	})
+	defer stop()
+
+	ng1, err := newNodegroupFromMachineSet(controller, machineSet1)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	ng2, err := newNodegroupFromMachineSet(controller, machineSet2)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	nodeNames1, err := ng1.Nodes()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	nodeNames2, err := ng2.Nodes()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if len(nodeNames1) != len(nodes)/2 {
+		t.Fatalf("expected len=%v, got len=%v", len(nodes)/2, len(nodeNames1))
+	}
+
+	if len(nodeNames2) != len(nodes)/2 {
+		t.Fatalf("expected len=%v, got len=%v", len(nodes)/2, len(nodeNames2))
+	}
+
+	sort.Strings(nodeNames1)
+	sort.Strings(nodeNames2)
+	sortedNodeNames := append(nodeNames1, nodeNames2...)
+
+	for i := 0; i < len(nodes); i++ {
```

Comment:
The same here, it will work for len(nodes) < 10

Reply:
The complete test is predicated on there only being 2 nodes.